### PR TITLE
refactor(LinearAlgeba/BilinearMap): Weaken `CommSemiring` to `SMulCommClass`

### PR DESCRIPTION
**Pull request by mans0954**
The results in the `CommSemiring` section of `LinearAlgeba/BilinearMap` don't require the semirings to be commutative. It is sufficient that scalar multiplication in the modules is commutative.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by mans0954** [comment]
Build failing due to:
```
Error: No space left on device : '/home/lean/actions-runner/_work/_actions/dcarbone/install-jq-action/v1.0.1'
```
Not something I think I can fix?

**Comment by ADedecker** [comment]
That's a thing that happens from time to time, the best you can do is ping maintainers on Zulip so that we can disable and fix the faulty runner. It's the case now so I've restarted the build for you.

File: Mathlib/LinearAlgebra/BilinearMap.lean [hunk 1]
```diff
@@ -385,7 +389,8 @@ variable (R M)
 /-- Scalar multiplication as a bilinear map `R → M → M`. -/
 def lsmul : R →ₗ[R] M →ₗ[R] M :=
   mk₂ R (· • ·) add_smul (fun _ _ _ => mul_smul _ _ _) smul_add fun r s m => by
-    simp only [smul_smul, smul_eq_mul, mul_comm]
+    simp only
+    rw [smul_comm]
```
Comment (ADedecker) [inline Mathlib/LinearAlgebra/BilinearMap.lean:393]:
```suggestion
  mk₂ R (· • ·) add_smul mul_smul smul_add fun r s => smul_comm s r
```

**Comment by ADedecker** [comment]
bors d+

File: Mathlib/LinearAlgebra/BilinearMap.lean (hunk 1 above)
Comment (mans0954) [reply Mathlib/LinearAlgebra/BilinearMap.lean]: Done, thanks!

**Review by eric-wieser — COMMENTED** [review]
What's the motivation for this change? Are there any interesting actions of non-commuting elements whose actions commute? I think in various places we avoided this generalization because there was no clear use case, and we expected a performance impact.

!bench

**Comment by mans0954** [comment]
> What's the motivation for this change? Are there any interesting actions of non-commuting elements whose actions commute? I think in various places we avoided this generalization because there was no clear use case, and we expected a performance impact.

@eric-wieser I can explain what I'm trying to do, and you can tell me if I'm going about it in the right way.

When I defined Jordan algebras, I [left out](https://github.com/leanprover-community/mathlib4/blob/master/Mathlib/Algebra/Jordan/Basic.lean#L41) [quadratic Jordan algebras](https://en.wikipedia.org/wiki/Quadratic_Jordan_algebra) which are more general as they work without restriction on the characteristic.

Now I've got a bit more used to working with Mathlib, I thought I'd see if I could give the quadratic definition, but that requires [quadratic maps](https://planetmath.org/quadraticmap). Currently, I think Mathlib only has [QuadraticForm](https://github.com/leanprover-community/mathlib4/blob/master/Mathlib/LinearAlgebra/QuadraticForm/Basic.lean#L130), but defined over a non-commutative ring.

I started working on quadratic forms over commutative rings, but it was quickly apparent that it's almost identical to the quadratic forms theory, and we would end up with two files with almost exactly the same results in them. I started thinking about whether it might be possible to develop the theory of quadratic maps over commutative rings and quadratic forms over non-commutative rings simultaneously, and I came up with:
```
structure QuadraticForm (R : Type u) (M : Type v) (N : Type w) [Semiring R] [AddCommMonoid M]
    [Module R M] [Module Rᵐᵒᵖ M] [AddCommMonoid N] [Module R N] [Module Rᵐᵒᵖ N] [SMulCommClass Rᵐᵒᵖ R N] where
  toFun : M → N
  toFun_smul : ∀ (a : R) (x : M), toFun (a • x) = (a * a) • toFun x
  toFun_smulr : ∀ (a : Rᵐᵒᵖ) (x : M), toFun (a • x) = (a * a) • toFun x
  exists_companion' : ∃ B : M →ₗ[R] M →ₗ[Rᵐᵒᵖ] N, ∀ x y, toFun (x + y) = toFun x + toFun y + B x y
```
The idea here being that `[SMulCommClass Rᵐᵒᵖ R R] ` will hold for an associative non-commutative ring as `R` is a left-module over `R` and a right-module over `Rᵐᵒᵖ`.

This seemed to be working, until I got to the [SMul](https://github.com/leanprover-community/mathlib4/blob/master/Mathlib/LinearAlgebra/QuadraticForm/Basic.lean#L365) section where I needed to define scalar multiplication on `QuadraticForm R M N` which requires `instance : SMul S (M →ₗ[R] M →ₗ[Rᵐᵒᵖ] N)` for the bilinear map.

You can see my work-in-progress branch here: https://github.com/leanprover-community/mathlib4/compare/master...mans0954/quadratic-maps

So I guess the options are:

- Merge this PR (if the performance is okay)
- Develop quadratic maps over commutative rings and quadratic forms over non-commutative rings separately
- Limit quadratic forms in Mathlib to only commutative rings
- Only have (linear) Jordan algebras in Mathlib?

**Comment by eric-wieser** [comment]
I'd envisaged that the correct statement was
```
toFun (a • x) = a • (op a • toFun x)
```
which corresponds to `Q(ax) = aQ(x)a`; does that sound right to you?

**Comment by eric-wieser** [comment]
Note that to replace the existing quadratic forms with your definition, we are first going to need #7152.

**Comment by eric-wieser** [comment]
Note that to replace the existing quadratic forms with your definition, we are first going to need #7152.

**Comment by mans0954** [comment]
> I'd envisaged that the correct statement was
> 
> ```
> toFun (a • x) = a • (op a • toFun x)
> ```
> 
> which corresponds to `Q(ax) = aQ(x)a`; does that sound right to you?

That looks plausible - I think the bilinear map is still `B : M →ₗ[R] M →ₗ[Rᵐᵒᵖ] N` and we will still need to show that `a • B` is also `M →ₗ[R] M →ₗ[Rᵐᵒᵖ] N`?

**Comment by eric-wieser** [comment]
Yes, I agree with the bilinear map form.

However, I don't think `r • B` is well-defined if `r : R` and `R` is not commutative. Better to define `s • B` for when the actions of `R` and `S` commute, which gives us the special case `r • B` for when `R` _is_ commutative.

**Comment by eric-wieser** [comment]
!bench

**Comment by mans0954** [comment]
> I'd envisaged that the correct statement was
> 
> ```
> toFun (a • x) = a • (op a • toFun x)
> ```
> 
> which corresponds to `Q(ax) = aQ(x)a`; does that sound right to you?

@eric-wieser  Do you also need:
```
toFun (op a • x) = op a • (a • toFun x)
```
i.e. `Q(xa) = aQ(x)a` e.g. to get `B(x,ya) = B(x,y)a`?

**Comment by eric-wieser** [comment]
No, that feels like a bad assumption to me as it means $$x \mapsto x^2$$ doesn't count as a quadratic form over a non-commutative ring

**Comment by mans0954** [comment]
> No, that feels like a bad assumption to me as it means x↦x2 doesn't count as a quadratic form over a non-commutative ring.
> 
> Edit: hmm, that seems troublesome anyway...

@eric-wieser I don't see that $$x \mapsto x^2$$  satisfies `Q(ax) = aQ(x)a` either?

**Comment by eric-wieser** [comment]
You're right, and that was what my edit was intended to convey.

I think the important question here is my earlier message:



> Do you have an example in mind (a choice of `R` and `M`) where `R` is non-commutative but its action on `M` is commutative?

**Comment by mans0954** [comment]
> I think the important question here is my earlier message:
> 
> > Do you have an example in mind (a choice of `R` and `M`) where `R` is non-commutative but its action on `M` is commutative?

Yes, sorry, I got a bit lost in all of this earlier and had my wires crossed. I agree now that I don't need this PR to get a scalar multiplication on `instance : SMul S (QuadraticForm R M N)` - I just need the appropriate `SMulCommClass` as hypothesis.

However, the next place I get stuck is in [composition of a quadratic map with a linear map](https://github.com/leanprover-community/mathlib4/pull/7569/files#diff-4b8a995fd51c4cb0a5d0ff6413fa1565f87f697d1b552327cfed2c98a4f62f49R601). Specifically `LinearMap.compl₁₂` doesn't work because it expects the rings to be commutative (https://github.com/leanprover-community/mathlib4/pull/7569/files#diff-4b8a995fd51c4cb0a5d0ff6413fa1565f87f697d1b552327cfed2c98a4f62f49R609). What I was hoping for from this PR was to have a version of `LinearMap.compl₁₂` that works with `SMulCommClass` instead of commutative rings.

I've opened a draft PR #7569 for the quadratic maps in case it's easier to discuss this there?

**Comment by eric-wieser** [comment]
> Specifically `LinearMap.compl₁₂` doesn't work because it expects the rings to be commutative

Well worse, doesn't it require the rings to be the same?

Note that in the short term you can use `((f ∘ₗ g).flip ∘ₗ h).flip)`; but it would be great to generalize `LinearMap.compl₁₂` to make it work here.

I think we should close this PR; I don't think there is anything worth merging here, and the conversation is still useful even if this PR is closed.

**Pull request closed without merging**